### PR TITLE
[Ray] Use Placement Groups for RayExecutor

### PR DESCRIPTION
**Pull request by amogkam**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by amogkam** [comment]
@richardliaw

File: horovod/ray/runner.py [hunk 1]
```diff
@@ -56,6 +58,10 @@ def hostname(self) -> str:
         # the intended hostname.
         return socket.gethostname()
 
+    def get_gpu_ids(self) -> List[int]:
+        """Return list of CUDA device Ids available to this worker."""
```
Comment (richardliaw) [inline horovod/ray/runner.py:62]:
```suggestion
        """Return list of CUDA device IDs available to this worker."""
```

File: horovod/ray/runner.py [hunk 2]
```diff
@@ -305,35 +234,71 @@ def __init__(self,
...
-        if not len(set(node_ids)) == len(node_ids):
-            raise RuntimeError("Colocator actors must "
-                               f"be placed on unique nodes! Got: {node_ids}")
+        bundles = [host_resources.copy() for _ in range(self.num_hosts)]
+        pg = ray.util.placement_group(bundles, strategy="STRICT_SPREAD")
+        self.placement_group = pg
+        logger.debug("Waiting for placement group to start.")
+        ready, _ = ray.wait([pg.ready()],
+                            timeout=int(os.environ.get(
+                                "HOROVOD_PLACEMENT_GROUP_TIMEOUT_S", 100)))
```
Comment (richardliaw) [inline horovod/ray/runner.py:251]: I think you can use the settings object for this instead of a separate env var

File: horovod/ray/runner.py [hunk 3]
```diff
@@ -305,35 +234,71 @@ def __init__(self,
...
+        self.workers = []
+        # Keep ref of one worker per node for NIC detection.
+        node_workers = []
+        # STRICT_SPREAD guarantees each bundle is on a different node.
+        # Create num_slots workers per bundle, i.e. per machine.
+        for bundle_index in range(len(bundles)):
+            gpu_id_futures = []
+            curr_node_workers = []
+            for i in range(self.num_slots):
+                remote_cls = ray.remote(BaseHorovodWorker)
```
Comment (richardliaw) [inline horovod/ray/runner.py:272]: can you move this `ray.remote` call outside of the forloop? It only needs to be called once

File: horovod/ray/runner.py (hunk 1 above)
Comment (amogkam) [reply horovod/ray/runner.py]: Done!

**Comment by richardliaw** [comment]
Generally looks fine:

1. could you read/follow the contributor guide to make the PR eligible for merge?
2. Could you also run `yapf` and `flake8` on `horovod/ray` (using the versions that we use on Ray core)  to maintain readability?

File: horovod/ray/runner.py (hunk 2 above)
Comment (amogkam) [reply horovod/ray/runner.py]: Good point, I added it as a new field to MiniSettings

File: horovod/ray/runner.py (hunk 3 above)
Comment (amogkam) [reply horovod/ray/runner.py]: yes good point! Fixed

File: horovod/ray/runner.py [hunk 4]
```diff
@@ -8,6 +8,8 @@
 from typing import Dict, Callable, Any, Optional, List
 import logging
 
+from ray.util import remove_placement_group
```
Comment (richardliaw) [inline horovod/ray/runner.py:11]: > WARNING: autodoc: failed to import module 'ray' from module 'horovod'; the following exception was raised:
No module named 'ray.util'

this line seems to break doc build; could you instead remove and just access it inline?

Comment (amogkam) [reply horovod/ray/runner.py]: yep, done

**Comment by EnricoMi** [comment]
I have cancelled the Buildkite build. Please rebase after #2814 has been merged into master.

File: horovod/ray/runner.py [hunk 5]
```diff
@@ -305,35 +241,74 @@ def __init__(self,
...
-        # We must save a pointer to each colocator to prevent their resource
-        # allocation from being released, along with their children from
-        # going out of scope.
-        self.colocators = colocators
-
-        node_ids = map_blocking(lambda a: a.create_workers.remote(),
-                                colocators)
-        if not len(set(node_ids)) == len(node_ids):
-            raise RuntimeError("Colocator actors must "
-                               f"be placed on unique nodes! Got: {node_ids}")
```
Comment (richardliaw) [inline horovod/ray/runner.py:332]: @yuduber FYI we are removing this runtime check in this PR

**Review by richardliaw — COMMENTED** [review]
This looks good, please ping me once tests pass.

**Comment by EnricoMi** [comment]
You can rebase to #2730 (branch-upgrade-test-versions-torch-1.8.0) to test your branch with torch 1.2.0 deprecated and rebase to master once that landed in master.

**Comment by tgaddair** [comment]
@amogkam looks like there are still some test failures, any thoughts on the cause?

https://buildkite.com/horovod/horovod/builds/5332#3176e70c-3fa6-4636-a052-7a32c55abe71/226-309

**Comment by amogkam** [comment]
@tgaddair yeah this is because of a bug on Ray 1.2. It's been fixed in nightly, but @richardliaw and I discussed this and we are planning to just hold off on this PR until the 1.3 release (by the end of the week) rather than use nightly Ray for the CI.

**Review by richardliaw — APPROVED** [review]
Looks good to me!

File: Dockerfile.test.cpu [hunk 6]
```diff
@@ -76,7 +76,7 @@ RUN if [[ ${SPARK_PACKAGE} != *"-preview"* ]]; then \
     fi
 
 # Install Ray.
-RUN pip install --no-cache-dir ray==1.2.0
+RUN pip install --no-cache-dir ray==1.3.0
```
Comment (richardliaw) [inline Dockerfile.test.cpu:79]: cc travis, as this PR should remove the flaky tests

**Comment by richardliaw** [comment]
Looks like tests are passing on buildkite but not on GHA. Do we need both?

**Comment by richardliaw** [comment]
@tgaddair I will merge this since:

1. It looks like the test 
> (Gloo Single PyTests (test-cpu-gloo-py3_7-tf1_15_5-keras2_2_4-torch1_3_1-mxnet1_5_1_p0-pyspark_3_1_1)
is just a bit flaky. The buildkite fully passes upon a retry.
2. The github actions is actually a bit misleading. It will indicate failure even if the test passes upon retry.

**Comment by richardliaw** [comment]
Screen Shot 2021-04-24 at 1 14 40 AM

**Pull request merged**